### PR TITLE
REST APIs that allows saving/edit/delete/search/share devfiles

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
This is a set of REST APIs that allows saving/edit/delete/search/share devfiles.
The code is marked as 'Preview Beta'. We expect some changes in future releases.
**Code can be reviewed commit by commit**

#### Model 
Main model object.
```json
{
  "name": "devfile-94c37pcwf49pcsy7",
  "namespace": "opentlc-mgr",
  "id": "id-8th8js34s6doo3q7",
  "devfile": {
    "metadata": {
      "name": "wksp-custom"
    },
    "apiVersion": "1.0.0"
  },
  "links": [
    {
      "href": "https://che-che.apps.cluster-f3d8.f3d8.example.opentlc.com/api/devfile/id-8th8js34s6doo3q7",
      "method": "GET",
      "rel": "self",
      "produces": "application/json"
    }
  ]
}
```
you can see such parts
| Field  | Description   |
|---|---|
|  id | Uniquely identifying persistent devfile   |
| name | Short human readable name of persistent devfile  | 
| description | A long explanation about how this concrete persistent devfile is used. | 
| namespace | Che namespace to which this persistent devfile belongs to | 
| devfile |Content of devfile that is matched Devfile schema 1.0 | 
| link | Weblinks related to this persistent devfile|

Such rest methods was implemented:
Знімок екрана 2020-09-16 о 10 58 08
-  As you can see there are two `POST` methods. One `devfile/devfile/` is used to create persistent devfile from yaml. The name would be generated randomly.
- For compatibility reasons method `GET /devfile` left untouched. It returns devfile 1.0 schema. Method `GET /devfile/list` is used to enumerate and search for devfiles.

## TODO for Phase 2
-  update CascadeRemovalTest, add userdevfiles cascade removal  https://github.com/eclipse/che/issues/17911
-  REST API to get by namespace `  @Path("/namespace/{namespace:.*}")`  https://github.com/eclipse/che/issues/17911
-  Replace get by ID with  get by composite key `  @Path("/{key:.*}")`. Composite key can be just devfile I D or in the  namespace:devfile_name form, where namespace is optional (e.g :devfile_name is valid key too. namespace/devfile_name form, where namespace can contain '/' character. https://github.com/eclipse/che/issues/17911

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16981 

### How to test this PR?
 - deploy che server image `quay.io/skabashn/che-server:che16981`
- run script  https://gist.github.com/skabashnyuk/97200e6b2e1df316e854d3d5fd1e326b

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java
Comment (sparkoo) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:254]: why accept only json? what about yaml ?

File: multiuser/permission/che-multiuser-permission-devfile/src/main/java/org/eclipse/che/multiuser/permission/devfile/server/spi/jpa/MultiuserJpaUserDevfileDao.java
Comment (sleshchenko) [inline multiuser/permission/che-multiuser-permission-devfile/src/main/java/org/eclipse/che/multiuser/permission/devfile/server/spi/jpa/MultiuserJpaUserDevfileDao.java]: it can be replaced with lamda

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java]: `s/referes on non-existent/refers to a non-existent/`

File: core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/UserDevfile.java
Comment (metlos) [inline core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/UserDevfile.java:27]: I know the "namespace" here means "user or organization" but I am not sure we should stick to this Che6 nomenclature. It is highly confusing to use the same name for completely different things in the model and in the infrastructures.

What about calling this `ownerName` or `accountName`? It would IMHO slightly better express where this is coming from as well as imply the same semantics (at least to an extent) - the same owner cannot have two devfiles with the same name, yet devfiles with the same name can have different owners.

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java]: `s/on name. But got:/by name but got:/`

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java]: `s/on name. But got:/by name or id but got:/`

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java]: `s/Possible values 'asc' or 'desc'. But got:/Possible values are 'asc' or 'desc' but got:`

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java:350]: We do the validation of the attributes upfront but maybe it should be noted somewhere that the `attribute.first` is put into the query further unsanitized. If we fail to remember that in the future and allow for "arbitrary" keys, we run in the risk of SQL injection.

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java]: Same thing as with `attribute.first` in the `withFilter` method. We should make sure (as much as comments allow) to not shoot ourselves in the foot in the future by allowing unsanitized string in the query.

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java]: We usually use constructor injection.

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java]: This is technically indenpendent of the Jpa dao. IMHO it would be better placed as a top level class instead of an inner class of the dao.

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/model/impl/UserDevfileImpl.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/model/impl/UserDevfileImpl.java]: Why the `+`?

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.19.0/1__userdevfile.sql
Comment (metlos) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.19.0/1__userdevfile.sql]: We're past `7.19.0`

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/UserDevfileManager.java
Comment (metlos) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/UserDevfileManager.java:72]: I think the user-devfile name is distinct from the name in the devfile on purpose. The former is the name of the "template" from which you create new workspaces using the devfile contained in it. So the name in the devfile really is a name of the workspace that is going to be created from it.

Comment (sparkoo) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/UserDevfileManager.java:72]: Yes, that's what I meant (maybe I wrote it in the way it's hard to understand). However, in current state of the PR, I think that you can't define name of the user-devfile when creating. So I'm suggesting here to add parameter on REST API creation method to allow caller to set the user-devfile name.

File: core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/UserDevfile.java
Comment (sparkoo) [reply core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/UserDevfile.java:27]: aha, that's what baffled me (https://github.com/eclipse/che/pull/17843#discussion_r490924937) :D

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.19.0/1__userdevfile.sql
Comment (sleshchenko) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.19.0/1__userdevfile.sql]: We're passing ) Master still has 7.19.0-SNAPSHOT

**Comment by skabashnyuk** [comment]
[crw-ci-test]

File: core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/UserDevfile.java
Comment (skabashnyuk) [reply core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/UserDevfile.java:15]: I was adding only files from this screenshot 
image from description
  https://github.com/eclipse/che/issues/16875

Comment (skabashnyuk) [reply core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/UserDevfile.java:15]: I've added https://github.com/eclipse/che/issues/17912 to add missing fields.  I propose to discuss there something that is missing.

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/UserDevfileManager.java
Comment (skabashnyuk) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/UserDevfileManager.java:72]: +1 for the extra parameter.  added as suggesting here https://github.com/eclipse/che/issues/17911

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java
Comment (skabashnyuk) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:254]: I was thinking about using yaml only for the createfromyaml method. Do you think someone will use yaml to update existed **user devfile**

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java
Comment (skabashnyuk) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:165]: Yes. In next issue https://github.com/eclipse/che/issues/17911

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java
Comment (sparkoo) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:152]: I see. API documentation is not clear about this. Does it make sense to have 2 methods? Do we have use-case for both?

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java
Comment (sparkoo) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:254]: I have no idea :) I'm used to see devfile in yaml so as I was playing with the API, it confused me a bit. But maybe it does not have a usecase. I don't know.

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java
Comment (sparkoo) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:133]: why not accept yaml here?

Comment (skabashnyuk) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:133]: I see no use-case for that. Do you?

**Comment by skabashnyuk** [comment]
@sparkoo @mshaposhnik @metlos @sleshchenko I think I've addressed your comments. Noticeable commits:
 - Applying reviewers suggestions https://github.com/eclipse/che/pull/17843/commits/d790930d3e02013ab4ad2596f4817b2d349cd4c1
- Applying reviewers suggestions https://github.com/eclipse/che/pull/17843/commits/50e9f0944af36947fdd80184ffa2a51b4616f362
- Added explanation of what UserDevfile.namespace is  https://github.com/eclipse/che/pull/17843/commits/f8cb092e165d05d6da6502c03062f09f539dc479
 -  Cover the possibility of using invalid search and order filed names https://github.com/eclipse/che/pull/17843/commits/a7f9eb30fb93b9f82ba6a31faaf573023b9cc92c

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java
Comment (sparkoo) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileService.java:133]: I don't know. Hard to imagine all use-cases. I guess we can add it in the future, if need raises

**Review by sparkoo — APPROVED** [review]

**Comment by skabashnyuk** [comment]
[crw-ci-test]

**Review by sleshchenko — COMMENTED** [review]
Sorry for the delay, I did not manage to redo the review in detail but I believe this PR should be good to merge once other reviewers approve it. So, feel free to merge without my approval )

File: multiuser/permission/che-multiuser-permission-devfile/src/main/java/org/eclipse/che/multiuser/permission/devfile/server/UserDevfileCreatorPermissionsProvider.java
Comment (mshaposhnik) [inline multiuser/permission/che-multiuser-permission-devfile/src/main/java/org/eclipse/che/multiuser/permission/devfile/server/UserDevfileCreatorPermissionsProvider.java]: unsubscribe ?

**Comment by skabashnyuk** [comment]
[ci-build]

**Comment by skabashnyuk** [comment]
[ci-build]

**Comment by skabashnyuk** [comment]
ci-build

**Review by mshaposhnik — APPROVED** [review]

**Comment by skabashnyuk** [comment]
ci-build

**Comment by skabashnyuk** [comment]
[ci-build]

**Comment by skabashnyuk** [comment]
[ci-build]

**Comment by skabashnyuk** [comment]
[ci-build]

**Pull request merged**
